### PR TITLE
plugins: initial support of parallel execution

### PR DESCRIPTION
**Pull request by nilslice**
This branches off of #57.

### CODE REVIEW

**Comment by nilslice** [comment]
@milton0825 - I think this branch is pretty close to done (aside from full integration tests with a plugin written in another language.. I may give that a shot and add it to the `plugin-samples` directory. 

Anyway, if you could give this a look and if possible run through the tests and/or try out a plugin that would be hugely helpful. 

I'll move on to your other PR in the meantime. Thanks again!

---
/cc @november-yankee, @rwong, @celrenheit (if you're also interested in taking a look!)

**Comment by milton0825** [comment]
@nilslice this is awesome! I can try build a plugin and share my experience back.

**Comment by milton0825** [comment]
I just tested it with my plugin and it works well. Great work!

File: extend/plugin.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	p.wrapErrAndLog(json.NewEncoder(os.Stdout).Encode(outputData))
+}
+
+func (p *plugin) wrapErrAndLog(err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stdout, "[protolock:plugin] %s: %v", p.name, err)
+	}
+}
+
+var _ Plugin = &plugin{}
```
Comment (milton0825) [inline extend/plugin.go]: Just curious. what is this for?

**Comment by nilslice** [comment]
> I just tested it with my plugin and it works well. Great work!

Thank you very much for testing! 

I will write up some documentation for this and try to find some time to add an example plugin in another language. Feel free to add any additional thoughts, concerns, issues, etc here.

File: extend/plugin.go (hunk 1 above)
Comment (milton0825) [reply extend/plugin.go]: ah I see. good learning

**Comment by milton0825** [comment]
cc @billonahill @astahlman @ArgentFalcon @rodaine

**Pull request merged**